plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p1cycles/DepthFirstCycleBreaker.java

```diff
@@ -0,0 +1,156 @@
+ * <ul>
+ * <li>Emden R. Gansner, Eleftherios Koutsofios, Stephen C. North, Kiem-Phong Vo, A technique for
+ * drawing directed graphs. <i>Software Engineering</i> 19(3), pp. 214-230, 1993.</li>
+ * </ul>
+ * 
+ * <p>This cycle breaker does not support the {@link LayeredOptions#PRIORITY_DIRECTION} option 
+ * that can be set on edges. Neither does it support layer constraints out of the box. 
+ * If layer constraints should be observed,
+ * {@link org.eclipse.elk.alg.layered.intermediate.EdgeAndLayerConstraintEdgeReverser} and
+ * {@link org.eclipse.elk.alg.layered.intermediate.LayerConstraintProcessor} should
+ * be used.</p>
+ * 
+ * <dl>
+ *   <dt>Precondition:</dt><dd>none</dd>
+ *   <dt>Postcondition:</dt><dd>the graph has no cycles</dd>
+ * </dl>
+ * 
+ * @see org.eclipse.elk.alg.layered.intermediate.EdgeAndLayerConstraintEdgeReverser
+ * @see org.eclipse.elk.alg.layered.intermediate.LayerConstraintProcessor
+ */
+public class DepthFirstCycleBreaker implements ILayoutPhase<LayeredPhases, LGraph> {
+
+    /** intermediate processing configuration. */
+    private static final LayoutProcessorConfiguration<LayeredPhases, LGraph> INTERMEDIATE_PROCESSING_CONFIGURATION =
+        LayoutProcessorConfiguration.<LayeredPhases, LGraph>create()
+            .addAfter(LayeredPhases.P5_EDGE_ROUTING, IntermediateProcessorStrategy.REVERSED_EDGE_RESTORER);
+
+    /** list of source nodes. */
+    private final List<LNode> sources = Lists.newArrayList();
+
+    /** mark for the nodes, inducing an ordering of the nodes. */
+    private int[] mark;
+    /** node at which the dfs started. */
+    private int[] root;
+    
+    /**
+     * {@inheritDoc}
+     */
+    public LayoutProcessorConfiguration<LayeredPhases, LGraph> getLayoutProcessorConfiguration(final LGraph graph) {
+        return INTERMEDIATE_PROCESSING_CONFIGURATION;
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void process(final LGraph graph, final IElkProgressMonitor monitor) {
+        monitor.begin("Depth-first cycle removal", 1);
+        
+        List<LNode> nodes = graph.getLayerlessNodes();
+
+        // initialize values for the algorithm 
+        int unprocessedNodeCount = nodes.size();
+        mark = new int[unprocessedNodeCount];
+        Arrays.fill(mark, -1);
+        root = new int[unprocessedNodeCount];
+        Arrays.fill(root, -1);
+        
+        int index = 0;
+        for (LNode node : nodes) {
+            // the node id is used as index for the indeg, outdeg, and mark arrays
+            node.id = index;
+            if (Iterables.isEmpty(node.getIncomingEdges())) {
+                sources.add(node);
+            }
+            index++;
+        }
+
+        // from every source node start a dfs
+        for (LNode source : sources) {
+            dfs(source, 0, source.id);
+        }
+        
+        // check if every node has been visited, otherwise start another dfs
```

Comment:
When can this happen? Once we have fiished DFS runs at all source nodes, shouldn't we have covered the graph, then?